src/hotspot/share/gc/g1/g1Allocator.cpp

```diff
@@ -458,9 +458,8 @@ HeapWord* G1ArchiveAllocator::archive_mem_allocate(size_t word_size) {
   // chunk if we don't yet cross the boundary, but the remainder would be too
   // small to fill.
   HeapWord* new_top = old_top + word_size;
```

Comment:
Pre-existing but related to the change - What, if anything, prevents the calculation of new_top from overflowing or detects that happened.  The usual way we do such calculations safely involves first checking that `(_max - old_top) <= word_size` or something similar.  I don't see anything to prevent that, and if nothing does then the various checks and calculations involving new_top are suspect.

Reply:
Thanks for pointing this out. A fix at https://github.com/openjdk/jdk/pull/3936

Reply:
I think the code is "good" but should be fixed - the `word_size` values are never user controlled afaict.